### PR TITLE
Allow mocking of final classes/methods on Android

### PR DESCRIPTION
**Pull request by moltmann**
by adding dexmaker-mockito-inline.

This is very similar to how mockito-inline works for JVMs but replaces
bytebuddy with a jvmti agent that uses the slicer lib.

If a class gets mocked for the first time the agent adds an entry hook
to all non-abstract methods. This entry hook calls into
MockMethodDispatcher->MockMethodAdvice which checks if the current
object is mocked. If so it executes
InvocationHandlerAdapter#doIntercept. The entryhooks are added as raw
DEX byte code in agent.cc.

Abstract methods are mocked via subclassing as in dexmaker-mockito. Of
course we don't have to mock non-abstract methods twice, hence we
restict the methods overridden by the subclasses.

This passes 99% of the mockito tests.
- Serialization is not supported yet
- Some mockito tests rely on bytebuddy and hence don't work on Android. We
  should abstract them out to be able to use dexmaker instead

I added some basic tests for mocking of final methods as
dexmaker-mockito-inline-tests. Further I fixed dexmaker-tests that were
failing on my current aosp-master build.

### CODE REVIEW

**Comment by moltmann** [comment]
It seems that we need to give Travis permission to use Android's API 27.

**Comment by moltmann** [comment]
Ok. Travis is happier now. We still need the NDK, but there is really no choice. Can we please add this to Travis?

File: dexmaker-mockito-inline/src/main/java/com/android/dx/mockito/inline/InlineDexmakerMockMaker.java [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+     * this class.
+     */
+    private static final Class DISPATCHER_CLASS;
+
+    /*
+     * One time setup to allow the system to mocking via this mock maker.
+     */
+    static {
+        // Allow to mock package private classes
+        System.setProperty("dexmaker.share_classloader", "true");
```
Comment (drewhannay) [inline dexmaker-mockito-inline/src/main/java/com/android/dx/mockito/inline/InlineDexmakerMockMaker.java]: Is this needed? The only reference I can find to it seems to be in Google's internal copy of dexmaker:
https://android.googlesource.com/platform/external/dexmaker/+/db20bbcc82de39f499a804d215851995041f3bcf%5E%21/#F0

**Comment by simon-the-canadian** [comment]
@drewhannay is this in 2.12.1? I'm trying to mock final classes and its not working. I also tried to add org.mockito.plugins.MockMaker and that doesn't work either.

**Comment by drewhannay** [comment]
@simon-the-canadian This won't work until Android P is released, and it also requires the new `dexmaker-inline` artifact that we haven't published yet (since Android P isn't out).

**Comment by simon-the-canadian** [comment]
Gotcha, thanks.

**Comment by jameswald** [comment]
@moltmann does this only allow mocking final elements on Android P? Is there any chance that this could eventually support older releases via additional contributions or is that unlikely due to inherent Android VM constraints on older platforms? Either way thanks for contributing this sorely needed improvement!

**Comment by moltmann** [comment]
Hi,

the approach I implemented requires APIs and improvements to ART's JVMTI
implementation that are only available in P.

Thanks
Philip

**Pull request merged**